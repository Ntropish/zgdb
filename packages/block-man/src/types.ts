export type Address = Uint8Array;

export interface BlockStore {
  /**
   * Retrieves a block of data from the store.
   * @param address The content address of the block.
   * @returns A Promise that resolves to the block's data as a Uint8Array, or undefined if not found.
   */
  get: (address: Address) => Promise<Uint8Array | undefined>;

  /**
   * Stores a block of data.
   * @param block The block data to store.
   * @returns A Promise that resolves to the content address of the stored block.
   */
  put: (block: Uint8Array) => Promise<Address>;

  /**
   * Checks for the existence of a block.
   * @param address The content address of the block.
   * @returns A Promise that resolves to true if the block exists, false otherwise.
   */
  has: (address: Address) => Promise<boolean>;
}

export interface Hasher {
  (data: Uint8Array): Promise<Address>;
}
