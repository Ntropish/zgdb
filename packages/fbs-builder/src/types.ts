export type FbsScalar =
  | "bool"
  | "byte"
  | "ubyte"
  | "short"
  | "ushort"
  | "int"
  | "uint"
  | "long"
  | "ulong"
  | "float"
  | "double";

export type FbsType = FbsScalar | string; // string for other tables, structs, etc.

export interface FbsField {
  name: string;
  type: FbsType;
  defaultValue?: number | string | boolean | null;
  isVector: boolean;
  attributes: Map<string, string | boolean>;
}

export interface FbsDeclaration {
  kind: "table" | "struct" | "enum" | "union";
  name: string;
  docs: string[];
}

export interface FbsTableState extends FbsDeclaration {
  kind: "table";
  fields: FbsField[];
  attributes: Map<string, string | boolean>;
}

export interface FbsStructState extends FbsDeclaration {
  kind: "struct";
  fields: FbsField[]; // in structs, fields cannot be optional or have defaults
}

export interface FbsEnumState extends FbsDeclaration {
  kind: "enum";
  type: FbsScalar;
  values: Map<string, number>;
}

export interface FbsUnionState extends FbsDeclaration {
  kind: "union";
  values: string[]; // table names
}

export interface FbsFileState {
  namespace: string | null;
  includes: string[];
  declarations: FbsDeclaration[];
  rootType: string | null;
  fileIdentifier: string | null;
  fileExtension: string | null;
}
