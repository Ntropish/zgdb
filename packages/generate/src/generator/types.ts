import { NormalizedSchema } from "../parser/types.js";

export interface GeneratorConfig {
  schemas: NormalizedSchema[];
  outputDirectory: string;
  options: {
    /**
     * The file extension to use for the generated import of the FlatBuffers schema.
     * E.g., '.js', '.mjs', or an empty string.
     * @default ".js"
     */
    importExtension?: string;
  };
}
