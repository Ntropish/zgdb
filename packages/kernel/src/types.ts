import { ComponentFactory } from "./component";

/**
 * A unique symbol used to signal that an orchestrator pipeline should halt execution.
 */
export const BREAK = Symbol.for("Orchestrator.BREAK");

/**
 * The most basic execution unit in the kernel. A function that accepts a context object
 * and can perform synchronous or asynchronous work.
 */
export type StepHandler<TContext extends Record<string, any> = any> = (
  ctx: TContext
) => unknown;

/**
 * The interface for a kernel that executes a linear pipeline of StepHandlers.
 */
export interface OrchestratorKernel<
  TContext extends Record<string, any> = any
> {
  run(ctx?: Partial<TContext>): Promise<TContext | typeof BREAK>;
  clone(): OrchestratorKernel<TContext>;
}

/**
 * A reactor's event handler is a self-contained orchestrator pipeline.
 */
export type EventHandler<TContext extends Record<string, any> = any> =
  OrchestratorKernel<TContext>;

/**
 * A structured collection of StepHandler arrays for logging at different levels.
 * This is designed to be injected into kernels to provide logging capabilities
 * without creating circular dependencies.
 */
export type LoggerPlugins = {
  info?: StepHandler[];
  warn?: StepHandler[];
  error?: StepHandler[];
};
