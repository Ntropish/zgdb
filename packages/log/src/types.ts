export type LogLevel = "fatal" | "error" | "warn" | "info" | "debug" | "trace";

export const LogLevels: Record<LogLevel, number> = {
  fatal: 60,
  error: 50,
  warn: 40,
  info: 30,
  debug: 20,
  trace: 10,
};

export interface LogRecord {
  level: LogLevel;
  message: string;
  time: number;
  data?: Record<string, unknown>;
  error?: Error;
}

export interface Transport {
  send(payload: string): void | Promise<void>;
}
