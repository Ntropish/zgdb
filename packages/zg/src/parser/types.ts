/**
 * This file defines the Intermediate Representation (IR) for our schema,
 * as well as the shape of the raw, user-defined schema files.
 */

import { z } from "zod";

// --- User-Facing Types ---

export type IndexDef = {
  on: string | string[];
  type?: "btree" | "hash" | "fulltext";
  unique?: boolean;
  description?: string;
};

export type PolymorphicRelationshipDef = {
  type: "polymorphic";
  cardinality: "one" | "many";
  required?: boolean;
  description?: string;
  discriminator: string; // e.g., 'targetType'
  foreignKey: string; // e.g., 'targetId'
  references: string[]; // e.g., ['Post', 'Comment']
};

export type StandardRelationshipDef = {
  type: "standard";
  entity: string;
  field?: string;
  cardinality: "one" | "many";
  required?: boolean;
  description?: string;
  mappedBy?: string;
};

export type RelationshipDef =
  | StandardRelationshipDef
  | PolymorphicRelationshipDef;

export type ManyToManyDef = {
  node: string;
  through: string;
  myKey: string;
  theirKey: string;
  description?: string;
};

export type EntityDef = {
  name: string;
  description?: string;
  schema: z.ZodObject<any>;
  relationships?: Record<string, RelationshipDef>;
  manyToMany?: Record<string, ManyToManyDef>;
  indexes?: IndexDef[];
  resolvers?: Record<string, Resolver<any>>;
};

export type ResolverContext<TClient = any, TActor = any> = {
  actor: TActor;
  db: TClient;
  input?: Record<string, any>;
  record?: Record<string, any>;
};

export type Resolver<TClient = any, TActor = any> = (
  context: ResolverContext<TClient, TActor>
) => Promise<boolean> | boolean;

export type Policy<TClient = any, TActor = any> = Resolver<TClient, TActor>;

export type InferredResolvers<
  TClient,
  TActor,
  TEntities extends Record<string, EntityDef>
> = {
  [K in keyof TEntities]?: Record<string, Resolver<TClient, TActor>>;
};

export type AuthBlock = {
  create?: string | string[];
  read?: string | string[];
  update?: string | string[];
  delete?: string | string[];
};

export type SchemaConfig<
  TClient = any,
  TActor = any,
  TEntities extends Record<string, EntityDef> = any,
  TGlobalResolvers extends Record<string, Resolver<TClient, TActor>> = any,
  TResolvers extends InferredResolvers<TClient, TActor, TEntities> = any
> = {
  entities: TEntities;
  globalResolvers?: TGlobalResolvers;
  resolvers?: TResolvers;
  auth?: {
    [entityName: string]: AuthBlock;
  };
};

// --- Internal, Normalized Types ---

export type Field = {
  name: string;
  type: string;
  attributes: Map<string, string | boolean>;
  description?: string;
  isVector?: boolean;
  required: boolean;
};

export type Index = {
  name: string;
  on: string[];
  unique?: boolean;
  type: "btree" | "hash" | "fulltext";
};

export type Relationship = {
  name: string;
  node: string;
  cardinality: "one" | "many";
  required?: boolean;
  description?: string;
  mappedBy?: string;
};

export type ManyToManyRelationship = {
  name: string;
  node: string;
  through: string;
  myKey: string;
  theirKey: string;
  description?: string;
};

export type PolymorphicRelationship = {
  type: "polymorphic";
  field: string;
};

export type NormalizedSchema = {
  name: string;
  description?: string;
  fields: Field[];
  isJoinTable?: boolean;
  indexes: Index[];
  auth?: AuthBlock;
  localResolvers?: Record<string, Function>;
  globalResolvers?: Record<string, Function>;
  relationships: (Relationship | PolymorphicRelationship)[];
  manyToMany: ManyToManyRelationship[];
};
