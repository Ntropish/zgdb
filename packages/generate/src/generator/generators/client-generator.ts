import { IGenerator } from "./interface.js";
import { NormalizedSchema } from "../../parser/types.js";

export class ClientGenerator implements IGenerator {
  generate(schemas: NormalizedSchema[]): string {
    const collectionProperties = schemas
      .filter((s) => !s.isJoinTable && !s.isNested)
      .map((s) => {
        const schemaName = s.name;
        const collectionName =
          schemaName.charAt(0).toLowerCase() + schemaName.slice(1) + "s";
        const collectionClassName = `${schemaName}Collection<TActor>`;
        return `  public readonly ${collectionName}: ${collectionClassName};`;
      })
      .join("\n");

    const constructorAssignments = schemas
      .filter((s) => !s.isJoinTable && !s.isNested)
      .map((s) => {
        const schemaName = s.name;
        const collectionName =
          schemaName.charAt(0).toLowerCase() + schemaName.slice(1) + "s";
        const collectionClassName = `${schemaName}Collection<TActor>`;
        const nodeName = `${schemaName}Node<TActor>`;
        const fbsNodeName = `${schemaName}FB.${schemaName}`;
        const getRootAs = `(bb) => ${fbsNodeName}.getRootAs${schemaName}(bb)`;
        const nodeFactory = `(tx, fbb: ${fbsNodeName}, ac) => new ${nodeName}(tx as ZgTransactionWithCollections<TActor>, fbb, ac)`;

        return `    this.${collectionName} = new ${collectionClassName}(this, '${schemaName}', ${nodeFactory}, ${getRootAs}, this.authContext);`;
      })
      .join("\n");

    const transactionClassName = `ZgTransactionWithCollections<TActor>`;

    return `
export class ${transactionClassName} extends ZgTransaction {
${collectionProperties}

  constructor(
    db: ZgDatabase,
    tree: any,
    authContext: ZgAuthContext<TActor> | null,
    config: ZgDbConfiguration,
  ) {
    super(db, tree, authContext, config);
${constructorAssignments}
  }
}

export const DB = {
  Transaction: ZgTransactionWithCollections,
};
`;
  }
}
