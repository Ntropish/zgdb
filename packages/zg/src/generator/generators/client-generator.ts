import { IGenerator } from "./interface.js";
import { NormalizedSchema } from "../../parser/types.js";

export class ClientGenerator implements IGenerator {
  generate(schemas: NormalizedSchema[]): string {
    const collections = schemas
      .filter((s) => !s.isJoinTable)
      .map((s) => {
        const schemaName = s.name;
        const collectionName =
          schemaName.charAt(0).toLowerCase() + schemaName.slice(1) + "s";
        const nodeName = `${schemaName}Node<TActor>`;
        return `  get ${collectionName}(): EntityCollection<${nodeName}> {
    return new EntityCollection<${nodeName}>(
      this,
      '${schemaName}',
      (db, fbb, ac) => new ${nodeName}(db, fbb, ac),
      (bb) => ${schemaName}FB.${schemaName}.getRootAs${schemaName}(bb)
    );
  }`;
      })
      .join("\n\n");

    return `export class ZgClient<TActor> extends ZgDatabase {
  constructor(
    prollyTree: ProllyTree,
    config: ZgConfig<TActor> = { entityResolvers: {}, globalResolvers: {} }
  ) {
    super(prollyTree, config);
  }

${collections}
}`;
  }
}
