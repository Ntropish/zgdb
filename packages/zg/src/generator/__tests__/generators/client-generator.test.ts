import { describe, it, expect } from "vitest";
import { ClientGenerator } from "../../generators/client-generator.js";
import { NormalizedSchema } from "../../../parser/types.js";

const testSchema: NormalizedSchema[] = [
  {
    name: "User",
    fields: [],
    indexes: [],
    relationships: [],
    manyToMany: [],
  },
  {
    name: "Post",
    fields: [],
    indexes: [],
    relationships: [],
    manyToMany: [],
  },
  {
    name: "Comment",
    fields: [],
    indexes: [],
    relationships: [],
    manyToMany: [],
    isJoinTable: true, // Should be ignored
  },
];

describe("ClientGenerator", () => {
  it("should generate a ZgClient class with collections for each schema", () => {
    const generator = new ClientGenerator();
    const result = generator.generate(testSchema);
    expect(result).toMatchSnapshot();
  });
});
