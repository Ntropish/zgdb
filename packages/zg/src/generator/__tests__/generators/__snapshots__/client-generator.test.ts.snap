// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ClientGenerator > should generate a ZgClient class with collections for each schema 1`] = `
"export class ZgClient<TActor> extends ZgDatabase {
  constructor(
    prollyTree: ProllyTree,
    config: ZgConfig<TActor> = { entityResolvers: {}, globalResolvers: {} }
  ) {
    super(prollyTree, config);
  }

  get users(): EntityCollection<UserNode<TActor>> {
    return new EntityCollection<UserNode<TActor>>(
      this,
      'User',
      (db, fbb, ac) => new UserNode<TActor>(db, fbb, ac),
      (bb) => UserFB.User.getRootAsUser(bb)
    );
  }

  get posts(): EntityCollection<PostNode<TActor>> {
    return new EntityCollection<PostNode<TActor>>(
      this,
      'Post',
      (db, fbb, ac) => new PostNode<TActor>(db, fbb, ac),
      (bb) => PostFB.Post.getRootAsPost(bb)
    );
  }
}"
`;
